traitsui/qt4/ui_panel.py

```diff
@@ -1300,6 +1324,95 @@ class TabbedFoldGroupEditor(GroupEditor):
+        if num_enabled_or_visible_whens > 0:
+            for object in self.ui.context.values():
+                object.on_trait_change(
+                    lambda: self._when(), dispatch="ui"
+                )
+            self._when()
+
+    def _when(self):
+        """Set the visible and enabled states of all tabs in the editor as
+        controlled by a 'visible_when' or 'enabled_when' expression.
+        """
+        self._evaluate_enabled_condition(self._enabled_when_groups)
+        self._evaluate_visible_condition(self._visible_when_groups)
+
+    def _evaluate_enabled_condition(self, conditions):
+        """Evaluates a list of (eval, widget) pairs and calls the
+        appropriate method on the widget to toggle whether it is
+        enabled as needed.
+        """
+        context = self.ui._get_context(self.ui.context)
+
+        if isinstance(self.container, QtGui.QTabWidget):
+            method_to_call_name = "setTabEnabled"
+        elif isinstance(self.container, QtGui.QToolBox):
+            method_to_call_name = "setItemEnabled"
+        else:
+            raise
+
+        for when, widget, idx in conditions:
+            method_to_call = getattr(self.container, method_to_call_name)
+            try:
+                cond_value = eval(when, globals(), context)
+                method_to_call(idx, cond_value)
+            except Exception:
+                # catch errors in the validate_when expression
+                from traitsui.api import raise_to_debug
+
+                raise_to_debug()
+
+    def _evaluate_visible_condition(self, conditions):
+        """Evaluates a list of (eval, widget) pairs and calls the
+        appropriate method on the tab widget to toggle whether it is
+        visible as needed.
+        """
+        context = self.ui._get_context(self.ui.context)
+
+        if isinstance(self.container, QtGui.QTabWidget):
+            tab_or_item = "Tab"
+        elif isinstance(self.container, QtGui.QToolBox):
+            tab_or_item = "Item"
+        else:
+            raise
+
+        for when, idx, widget, label in conditions:
+
+            try:
+                cond_value = eval(when, globals(), context)
+                if cond_value:
+                    method_to_call_name = "insert" + tab_or_item
+                    method_to_call = getattr(
+                        self.container, method_to_call_name
+                    )
+                    # check that the tab for this widget is not already showing
+                    if self.container.indexOf(widget) == -1:
+                        method_to_call(idx, widget, label)
+                else:
+                    method_to_call_name = "remove" + tab_or_item
+                    method_to_call = getattr(
+                        self.container, method_to_call_name
+                    )
+                    # check that the tab for this widget is already showing
+                    if self.container.indexOf(widget) != -1:
```

Comment:
theoretically, if the widget is visible, `idx == self.container.indexOf(widget)`.  However, perhaps the tab could change its index during the lifetime of the tab widget, in which case we should call `method_to_call(self.container.indexOf(widget))` instead?

Reply:
I think that this comment raises a significant issue that needs to be considered: the index of the widget in the `Tabbed` group doesn't necessarily match the index of the Qt tab, because visibility is controlled by adding/removing the widget from the container.

I think things will go wrong in situations like the following:
- have tabs T0, T1, T2, T3, all initially visible
- T1 is hidden, so we get T0, T2, T3, but now the indices of T2 and T3 at the Qt level are 1 and 2, but at the TraitsUI level are 2 and 3.
- If we hide T2 in this state, the current code will actually hide T3 (and hiding T3 will result in an error?)
- After hiding T2 we have T0 and T2 at the Qt level.  If we un-hide T2, then it will try to insert it at Qt index 2 with the current code, and so we would get no change as it is already showing

I think there are situations where you can get things showing up in pretty much arbitrary order (eg. I think hide T2, hide T1, show T2 ends up with T0, T3, T2 as the layout).

I could be missing something about indices are being computed, but it looks like you can't assume that TraitsUI and Qt-level indices match, and computing the right Qt level index for re-insertion needs some thought (I think we need to work out how many TraitsUI tabs before the tab of interest are visible).